gallery/gallery/lib/studies/shrine/supplemental/balanced_layout.dart

```diff
@@ -0,0 +1,305 @@
+// Copyright 2019 The Flutter team. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import 'package:gallery/studies/shrine/model/product.dart';
+import 'package:gallery/studies/shrine/supplemental/desktop_product_columns.dart';
+import 'package:gallery/studies/shrine/supplemental/layout_cache.dart';
+
+/// A placeholder id for an empty element. See [_iterateUntilBalanced]
+/// for more information.
+const _emptyElement = -1;
+
+/// To avoid infinite loops, improvements to the layout are only performed
+/// when a column's height changes by more than
+/// [_deviationImprovementThreshold] pixels.
+const _deviationImprovementThreshold = 10;
+
+/// Height of a product image, paired with the product's id.
+class _TaggedHeightData {
+  const _TaggedHeightData({
+    @required this.index,
+    @required this.height,
+  });
+
+  /// The id of the corresponding product.
+  final int index;
+
+  /// The height of the product image.
+  final double height;
+}
+
+/// Converts a set of [_TaggedHeightData] elements to a list,
+/// and add an empty element.
+/// Used for iteration.
+List<_TaggedHeightData> toListAndAddEmpty(Set<_TaggedHeightData> set) {
+  List<_TaggedHeightData> result = List<_TaggedHeightData>.from(set);
+  result.add(_TaggedHeightData(index: _emptyElement, height: 0));
+  return result;
+}
+
+/// Sort a list of integers, then return the sorted list.
+/// Used in list comprehensions using "for".
+List<int> _sorted(List<int> list) {
+  list.sort();
+  return list;
+}
+
+/// Encode parameters for caching.
+String _encodeParameters({
+  @required int columnCount,
+  @required List<Product> products,
+  @required double largeImageWidth,
+  @required double smallImageWidth,
+}) {
+  final String productString =
+      [for (final product in products) product.id.toString()].join(',');
+  return '$columnCount;$productString,$largeImageWidth,$smallImageWidth';
+}
+
+/// Given a layout, replace integers by their corresponding products.
+List<List<Product>> _generateLayout({
+  @required List<Product> products,
+  @required List<List<int>> layout,
+}) {
+  return [
```

Comment:
nit: Are you by any chance a skilled python programmer? :)

This is more of a style FYI than anything else, but typically these types of list manipulations use `map` and Dart's other iterator-based methods. No need to rewrite anything here, I just wanted to mention it.

Reply:
Thank you very much for the suggestion!
I'll keep it like this for this PR, but I'll keep this in mind for future PRs.
I do use python a lot, but I'm still learning.      :D